modules/basics/src/main/java/com/opengamma/strata/basics/schedule/PeriodicSchedule.java

```diff
@@ -1060,6 +1063,54 @@ public AdjustableDate calculatedEndDate() {
     return AdjustableDate.of(endDate, calculatedEndDateBusinessDayAdjustment());
   }
 
+  //-------------------------------------------------------------------------
+  /**
+   * Returns an instance based on this schedule with the start date replaced.
+   * <p>
+   * This returns a new instance with the schedule altered to have the specified start date.
+   * The specified date is considered to be adjusted, thus {@code startDateBusinessDayAdjustment} is set to 'None'.
+   * The {@code firstRegularStartDate} and {@code overrideStartDate} fields are also removed.
+   * <p>
+   * The stub convention is typically altered to be 'SmartInitial'.
+   * The algorithm retains the 'ShortInitial' and 'LongInitial' conventions as is.
+   * The algorithm examines "Final" conventions to try and set the {@code lastRegularEndDate}
+   * via schedule generation allowing the stub convention to become 'SmartInitial'.
+   * 
+   * @param adjustedStartDate the proposed start date, which is considered to be adjusted
+   * @return a schedule with the proposed start date
+   * @throws IllegalArgumentException if the schedule start date cannot be replaced with the proposed start date
+   */
+  public PeriodicSchedule replaceStartDate(LocalDate adjustedStartDate) {
+    if (adjustedStartDate.isAfter(endDate)) {
+      throw new IllegalArgumentException("Cannot alter leg to have start date after end date");
+    }
+    PeriodicSchedule.Builder builder = toBuilder()
+        .startDate(adjustedStartDate)
+        .startDateBusinessDayAdjustment(BusinessDayAdjustment.NONE)
+        .firstRegularStartDate(null)
+        .overrideStartDate(null);
+    if (stubConvention == null || stubConvention == StubConvention.BOTH || stubConvention == StubConvention.NONE) {
+      // set the stub convention to SmartInitial to better handle the new start date
+      builder.stubConvention(StubConvention.SMART_INITIAL);
+    } else if (stubConvention.isFinal()) {
+      if (lastRegularEndDate != null) {
+        // last regular is set, so the final stub convention can be safely changed
```

Comment:
I don't follow why the stub convention can be safely changed in this case. The roll day of the last regular end date should be in sync with the roll day of the trade start date.

For example, if I have existing trade with
Start Date: 2021-06-01
End Date: 2026-07-01
Frequency: 6M
Last Regular End Date: 2026-06-01

This gives me a trade with 6M periods and a 1M stub at the end.

If I then come along and change the start date to 21-06-15 using this method I am left with what looks like an inconsistent trade definition
Start Date: 2021-06-15
End Date: 2026-07-01
Frequency: 6M
Last Regular End Date: 2026-06-01
Stub Convention: Smart Initial

So the presence of a Last Regular End date that is different to the trade end date implies a back stub. 
But we are then explicitly saying this trade has an initial stub

Reply:
Smart initial has been changed to cope with this. The smart stub conventions now respect any first or last regular date, and apply themselves to the remaining period.

Reply:
Ok so `SmartInitial` can actually result in a trade with both a front and back stub?

Not the most intuitive but we're in to real edge case territory here anyway, so I agree the behaviour here is sensible enough

Reply:
Yes, `SmartInitial` and `SmartFinal` now have reasonable behavour for a use case that previously threw an exception.